File: decorator/Decorator.java
public class Decorator {
    private interface IWeapon {
        void describe();

        int getAttack();
    }

    private class Sword implements IWeapon {

        @Override
        public void describe() {
            System.out.println("This is a sword");
        }

        @Override
        public int getAttack() {
            return 10;
        }
    }

    private interface IWeaponDecorator extends IWeapon {

    }

    private class SuperSwordDecarator implements IWeaponDecorator {

        protected IWeapon weapon;

        public SuperSwordDecarator(IWeapon weapon) {
            this.weapon = weapon;
        }

        @Override
        public void describe() {
            System.out.println("This is an upgraded");
            weapon.describe();
        }

        @Override
        public int getAttack() {
            return weapon.getAttack() + 5;
        }
    }

    public void testDecorator() {
        Sword sword = new Sword();
        // normal sword
        sword.describe();
        System.out.println("Attack: " + sword.getAttack() + "\n");

        // super sword
        SuperSwordDecarator superSword = new SuperSwordDecarator(sword);
        superSword.describe();
        System.out.println("Attack: " + superSword.getAttack());
    }

    public static void main(String[] args) {
        Decorator decorator = new Decorator();
        decorator.testDecorator();
    }
}

File: decorator/Decorator.cpp
#include <iostream>
#include <string>

using namespace std;

// abstract class weapon interface
class IWeapon
{
public:
    virtual void describe() = 0;
    virtual int getAttack() = 0;
};

class Sword : public IWeapon
{
public:
    void describe()
    {
        cout << "This is a sword" << endl;
    };
    int getAttack()
    {
        return 10;
    };
};

class IWeaponDecorator : public IWeapon
{
};

class SuperSwordDecorator : public IWeaponDecorator
{
private:
    IWeapon *weapon;

public:
    SuperSwordDecorator(IWeapon *weapon)
    {
        this->weapon = weapon;
    }
    void describe()
    {
        cout << "This is an upgraded" << endl;
        weapon->describe();
    };
    int getAttack()
    {
        return 20 + weapon->getAttack();
    };
};

void testDecorator()
{
    IWeapon *sword = new Sword();
    sword->describe();
    cout << "Attack: " << sword->getAttack() << endl
         << endl;

    IWeapon *superSword = new SuperSwordDecorator(sword);
    superSword->describe();
    cout << "Attack: " << superSword->getAttack() << endl;
}

// as a client
int main(int argc, char *argv[])
{
    testDecorator();
    return 0;
}